zingo-price/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+#![warn(missing_docs)]
+
+//! Crate for fetching historical and live ZEC prices
+
+use std::io::{Read, Write};
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use serde::Deserialize;
+
+use zcash_encoding::{Optional, Vector};
+
+#[derive(Debug, Deserialize)]
+struct HistoricalPriceData {
+    /// Date.
+    #[serde(rename = "date")]
+    _date: String,
+    /// ZEC price in USD.
+    #[serde(rename = "priceUsd")]
+    price_usd: String,
+    /// Time in milliseconds.
+    time: u128,
+}
+
+#[derive(Debug, Deserialize)]
+struct CurrentPriceData {
+    /// ZEC price in USD.
+    #[serde(rename = "priceUsd")]
+    price_usd: String,
+}
+
+/// Price of ZEC in USD at a given point in time.
+#[derive(Debug, Clone, Copy)]
+pub struct Price {
+    /// Time in seconds.
+    pub time: u32,
+    /// ZEC price in USD.
+    pub price_usd: f32,
+}
+
+/// Price list for wallets to maintain an updated list of daily ZEC prices.
+#[derive(Debug)]
+pub struct PriceList {
+    /// Time of last price update in seconds.
+    time_last_updated: Option<u32>,
+    /// Current price.
+    current_price: Option<Price>,
+    /// Historical price data by day.
+    daily_prices: Vec<Price>,
+}
+
+impl PriceList {
+    /// Constructs a new price list from the time of wallet creation.
+    pub fn new() -> Self {
+        PriceList {
+            time_last_updated: None,
+            current_price: None,
+            daily_prices: Vec::new(),
+        }
+    }
+
+    /// Returns time price list was last updated.
+    pub fn time_last_updated(&self) -> Option<u32> {
+        self.time_last_updated
+    }
+
+    /// Returns current price.
+    pub fn current_price(&self) -> Option<Price> {
+        self.current_price
+    }
+
+    /// Returns historical price data by day.
+    pub fn daily_prices(&self) -> &[Price] {
+        &self.daily_prices
+    }
+
+    /// Price list requires a start time before it can be updated.
+    ///
+    /// Recommended start time is the time the wallet's birthday block height was mined.
+    pub fn set_start_time(&mut self, time_of_birthday: u32) {
+        self.time_last_updated = Some(time_of_birthday);
+    }
+
+    /// Updates price list.
+    pub async fn update(&mut self) -> Result<(), PriceError> {
+        // FIXME: move to secret or user specified
+        let api_key = "4bce48cf8766d5c55ecdd83622cbba676fdc23745b7176916fd517b40e5ee6d0";
```

Comment:
Who are we registering zingo with?

Reply:
zingo is not registering, its up to the user. this fixed key was for testing